src/lib.rs

```diff
@@ -138,26 +223,30 @@ where
   zn_secondary: G2::Scalar,
   _p_c1: PhantomData<C1>,
   _p_c2: PhantomData<C2>,
+  a1: PhantomData<A1>,
+  a2: PhantomData<A2>,
+  num_steps: usize,
 }
 
-impl<G1, G2, C1, C2> RecursiveSNARK<G1, G2, C1, C2>
+impl<G1, G2, C1, C2, A1, A2> RecursiveSNARK<G1, G2, C1, C2, A1, A2>
 where
   G1: Group<Base = <G2 as Group>::Scalar>,
   G2: Group<Base = <G1 as Group>::Scalar>,
-  C1: StepCircuit<G1::Scalar> + Clone,
-  C2: StepCircuit<G2::Scalar> + Clone,
+  C1: StepCircuit<G1::Scalar, A1> + Clone,
+  C2: StepCircuit<G2::Scalar, A2> + Clone,
+  A1: Arity<<G1 as Group>::Scalar>,
+  A2: Arity<<G2 as Group>::Scalar>,
 {
-  /// Create a new `RecursiveSNARK`
-  pub fn prove(
-    pp: &PublicParams<G1, G2, C1, C2>,
-    num_steps: usize,
+  /// Initialize a `RecursiveSNARK`.
+  pub fn init(
+    pp: &PublicParams<G1, G2, C1, C2, A1, A2>,
     z0_primary: G1::Scalar,
     z0_secondary: G2::Scalar,
+    z1_primary: G1::Scalar,
+    z1_secondary: G2::Scalar,
+    first_primary_circuit: &C1,
+    first_secondary_circuit: &C2,
   ) -> Result<Self, NovaError> {
-    if num_steps == 0 {
```

Comment:
How do we handle this case without this code in the new version?

Reply:
The error will be returned if an iterator has no first item:
```
https://github.com/microsoft/Nova/pull/76/files/3986df4f2c580d49f449f84ab84c02eb0db1668b#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R394-R401
```